### ISSUE TITLE
explore https://gioui.org/ as UI

### ISSUE THREAD

**Issue by cbrake** [simpleiot/simpleiot#270]
@gedw99 lets start a new issue for this exploration

Discussion of nats.go/wasm: https://github.com/nats-io/nats.go/issues/831

**Comment by gedw99** [comment]
thanks for chiming in about this.

I wanted to find out the situation with NATS support of WASM, and well know we know. Its 2020 Q2 IF we are lucky.

So when looking at the https://pkg.go.dev/nhooyr.io/websocket, i noticed that tons of others have used it to make a NATS proxy :) Seems that this might be the way to code.

I am still not discounting SSE though, If we have to build a proxy for NATS, then why not just build a SSE proxy and get all the good things of SSE ( native browser support), and a very simple protocol that can easily be mapped to via a Proxy back to NATS. the IOT devices just bypass the proxy and go direct to NATS.

I am going to research both options. But if you have and thoughts of SSE versus WS, please yell :)

**Comment by cbrake** [comment]
yeah, I don't know enough about either web sockets or SSE to have a strong opinion.

With SSE, would you also have to proxy at both the frontend and the backend?

Another consideration is AuthZ/N -- I'm looking at leveraging NATS to handle all the Auth issues, so would need to consider all that if we proxy anything.

**Comment by cbrake** [comment]
Here is the brainstorming on Auth: https://github.com/simpleiot/simpleiot/issues/268

**Comment by gedw99** [comment]
Regarding the web sockets and gio ..

 i think i have found a good clean solution for Web sockets talking from the GIO gui to the golang server.

https://git.sr.ht/~whereswaldon/pointstar

Its a websocket client and server for gio. The code is showing it off as a game, and it that keeps state as a golang struct, just to keep things simple.
The code is very clean and minimal.
This will means that the GUI can run on Web, Desktop or Mobile. 
You can try both with this code as it pulls in the gogio command for you and compiles to wasm, but also to desktop.

it uses nhooyr.io/websocket which uses https://github.com/gobwas/ws
here is a nice  write up by gobwas developer the of its scalable design:  https://sergey.kamardin.org/articles/million-websocket-and-go/

So i think we can easily hook this up to NATS on the server.
Also because its a web socket, its patterns of communication are similar  to how the IOT patterns of communication.  Basically the data flow patterns are similar.

SSE is not viable i felt because of its lack of binary support.

I have not had much time to play with Chris's code, as i just wanted to explain that this is a good option.
Am now working on i18n aspects.


---

Also we need a Tree control for gio to replace the current elm one.

Chris has one it seems:  https://gophers.slack.com/archives/CM87SNCGM/p1632402647359900
- he will break it out of his code and publish it open source soon.

Its looks nice and flexible. It seems to be being used on both the left and right panes. 
![gio-tree-image](https://user-images.githubusercontent.com/53147028/134694176-0c958470-9097-4a95-baa7-382ece24002c.png)

--- 

Testing....

I was thinking that we can use web page with 4 iframes in it.
top 2 run current elm gui.
bottom 2 run gio wams gui.
left and right are used to express "root" and "org A" access

I got this idea from watching your video , and realised a nice simple Iframe will make developing and testing easier.

For Desktop. I think its better to use websockets and and not nats. why ?
- cause we have parity between web, desktop and mobile
- more secure to hid nats on server behind Websockets.

---

I also played with the idea of using a libp2p API so that we can do Client <--> Server, but also Client <--> Client ( p2p ).
There are some use cases for this i personally have. But anyway the above approach will get things cooking for now...

**Comment by cbrake** [comment]
interesting stuff. I tried to build pointstart:

```
[cbrake@ceres pointstar]$ make
go build -o pointstar-server ./server
go: downloading nhooyr.io/websocket v1.8.4
go: downloading github.com/klauspost/compress v1.10.0
go build -o pointstar ./client
go: downloading gioui.org v0.0.0-20200510064036-c2cbcee78de0
go: downloading golang.org/x/image v0.0.0-20190802002840-cff245a6509b
go: downloading golang.org/x/sys v0.0.0-20200124204421-9fbb57f87de9
# gioui.org/app/internal/egl
/home/cbrake/go/pkg/mod/gioui.org@v0.0.0-20200510064036-c2cbcee78de0/app/internal/egl/egl_unix.go:39:3: cannot use nil as type _Ctype_EGLConfig in return argument
make: *** [makefile:12: client] Error 2
```

Looking at code:

![image](https://user-images.githubusercontent.com/402813/134698196-db89be99-a730-46cf-976a-9f7411d3673e.png)

I then tried a gio kitchen example, that runs fine:

![image](https://user-images.githubusercontent.com/402813/134698613-9d25d165-7a5b-4c99-aa40-ce4230597822.png)

**Comment by gedw99** [comment]
wow your fast !!

Hate to say the old "works for me", but it did.

 I am on "go version go1.17.1 darwin/amd64".
 Its family dinner time here, so cant keep going on this today...  Happy to grok it tomorrow if you have no luck.



My build using his makefile:
```
make all
go build -o pointstar-server ./server
go build -o pointstar ./client
rm -rf pointstar-assets/ && go run gioui.org/cmd/gogio -target js -o pointstar-assets ./client
./pointstar-server
```

Your error maybe be because his code is importing on an old gio version that had some linux issues.
upgrade the go mod.

this makes it easy:
``` 
go get github.com/oligot/go-mod-upgrade
go-mod-upgrade
go mod tidy
go mod verify
```

After i upgraded all the go packages in one hit . Get an error here:
``` 
make all
go build -o pointstar-server ./server
go build -o pointstar ./client
# git.sr.ht/~whereswaldon/pointstar/client/widgets
client/widgets/card.go:12:2: undefined: widget.Button
``` 

Also here is a basic golang file server to run the wasm:
``` 
package main

import (
	"log"
	"net/http"
)

func main() {

	http.Handle("/", http.FileServer(http.Dir(".")))

	log.Println("Serving from 8081")

	log.Fatal(http.ListenAndServe(":8081", nil))
}
```

Tomorrow i will fork and work on it.

**Comment by gedw99** [comment]
Whoops your on Linux. Did you I stall the OS dependencies that gio needs for Linux ?

https://gioui.org/doc/install#linux

**Comment by cbrake** [comment]
I'm on Arch Linux. I assume since the gio demo works, I have needed dependencies.

**Comment by gedw99** [comment]
> I'm on Arch Linux. I assume since the gio demo works, I have needed dependencies.

true. i forgot you got kitchen working..

**Comment by cbrake** [comment]
This illustrates one fundamental truth -- any time you get C dependencies involved, you are up against build problems. May not be your build, but someone will hit build problems. I've spent the last decade trying to tame embedded Linux build problems with mostly C code:

https://github.com/YoeDistro/yoe-distro/graphs/contributors

It has provided a good living (no one else wants to do this kind of stuff), but I'm convinced there is a better way. We don't fix problems by piling on layer after layer of complexity. From this perspective, pure Go is the development experience I've been looking for, and I feel is the best path forward for anybody developing embedded Linux apps (my customers). 

There are tradeoffs -- gioui may be compelling enough to live with the CGO build issues. Maybe WASM will be CGO free. I feel that the future (at least for industrial embedded Linux apps) is browser based UIs in some form or another.

**Comment by gedw99** [comment]
I hear ya . There is a ton of work sorting out stuff others hate.

i have found gio to be really great and gio wasm is cgo free.

All of gio btw is cgo free. It’s designed specifically to not use cho because of the  pain you slides to and other reasons. You can raise a discussion on the mailing list as I am sure you have more knowledge about this than me.

If you want to just stick with elm let me know - was not sure from your text how you felt. I recommend trying gio at least.

**Comment by cbrake** [comment]
That is great gio wasm is CGO free -- certainly holds promise.

I really like Elm for the following reasons:

- has the same values as Go -- simplicity
- functional programming is neat
- produces very high quality results (no run time exceptions in the browser)
- community is great, friendly, helpful, etc
- elm-ui is amazing
- reasonably popular: https://github.com/jah2488/elm-companies
- the hard work is done (rending node tree, etc). Adding new features at this point is relatively easy.
- asset size is tiny (https://community.tmpdir.org/t/comparison-of-asset-sizes-for-various-front-end-technologies/213/3) -- feel free to contribute to this tread if you have go/wasm info.

That said, there are some disadvantages of Elm:

- its not Go
- its functional and Go is not
- learning curve is much steeper than Go

So I realize for OSS adoption, Elm is likely a hindrance on this project.

Exploring alternatives is a good idea. SIOT can certainly have two or more UIs. The architecture (everything is done through NATs) makes it easy to bolt on additional modules/UIs that are not part of the core application. For example, you could write a machine learning module in python that connects via NATs and processes data and generates new data.

So I'm all for exploring alternatives -- just want to make sure we discuss tradeoffs, project goals, etc.

Can you help understand how gio is CGO free? In the `go/pkg/mod/gioui.org@v0.0.0-20200510064036-c2cbcee78de0/app/internal/egl/egl_unix.go` file above, it seems there is CGO stuff.

**Comment by gedw99** [comment]
About CGO, i only meant for WASM. For Desktop and Mobile, yes it has to do some CGO calls.
https://github.com/gioui/gio/blob/main/internal/cocoainit/cocoa_darwin.go

---

About tradeoffs and goals... My 2 cents:

One of my goals is to build GUI apps on top of SIOT. Web, Desktop and Mobile. I find GIO to be the best for cross platform goal. 

I see SIOT as a backend, and the SIOT admin GUI as a thing used by the dev / ops person only.  

But both the User GUI and the Admin GUI need a push / pull communication mechanism, and so i started leaning to GIO so that i had to only deal with one gui style and hence one way to do the push pull.

https://github.com/centrifugal/centrifuge is another option. 
Good:
- Its nice in that is has many options, and is decoupled from the GUI and the backend. 
- can can as lib or a server. 
- can connect to nats. https://github.com/centrifugal/centrifugo/blob/master/internal/natsbroker/broker.go
- can use bidirectional or unidirectional communication patterns.  https://centrifugal.dev/docs/transports/overview
- Been though battle testing. There are alot of gotchas with async communications systems, and its been through a fair bit.
- has presence. But ONYL for the bidirectional
- is language agnostic. Can use grpc etc.
Bad:
- no GIO or elm example that stands out.

**Comment by gedw99** [comment]
https://user-images.githubusercontent.com/53147028/134799833-d303d3dc-640f-4d50-a9fa-f1e3ecaf782b.mov

**Comment by cbrake** [comment]
I'm interesting in seeing how the GIO use case might work out. Would probably be simplest to start with a desktop app and see if you can connect via NATS and display a node tree, or some useful demo. Looking forward to seeing how it works out!

first I've heard of centrifuge -- you are a library of OSS technology! -- looks neat, but I'd need to understand what specific things it will do that we can't do with NATS. Again, my approach is solve real-world problems and add/change technology as I go -- not add in boatloads of dependencies at the start. A simple single binary deployment is also very important to me (especially at the edge). The symmetry of running the same SIOT apps at the edge and in the cloud is also a very powerful. However, it also requires a focus on simplicity because resources at the edge (embedded Linux systems) is limited.

**Comment by gedw99** [comment]
I agree about centrifuge. Before Pointstar was validated, it might have been needed. Its not at all needed now :)
Now its just a matter of hooking up Pointstar to NATS and your basically done. NATS is stateful and can push messages as needed up through Pointstar and visa versa. Very simple. Single binary as you said.. 

Also Form widget to go with the tree gio widget. 
Its very simple to use: https://git.sr.ht/~jackmordaunt/gio-planet/tree
Kind of amazed this has not been added to the GIO examples yet, cause its such a standard thing.

Also i recommend you to run the gio component as the basis of the real app.
https://github.com/gioui/gio-example/tree/main/component
It has all the usuals things that are needed to round out an app.
then you can add the Tree and Form widgets.

I think give or take a few things, that basically gives the basis to use GIO instead of Elm if you want to.

**Comment by gedw99** [comment]
ok about web verses desktop / mobile.

I would use web sockets for all 3 because then you have same code everywhere.  Just simpler for consistency. But that’s maybe just my bias.

for iot you have no gui and so then nats is perfect. 

The point is that the web socket layer is stateless . That is the most important aspect.

If you want me to look at something just shout. I will try to also work on this repo but don’t have a ton of time this week.

**Comment by gedw99** [comment]
I wish i had time to work on this, but am slammed with work...

So, what i mean about the WS and NATS, is that the WS is doing the network transport and connection handing only between the GUI and the Server.

So with the https://git.sr.ht/~whereswaldon/pointstar demo, i imagine that all that is needed, is to make it talk to NATS on the Server.

So i did a quick search of github to find examples of this: https://github.com/search?q=nhooyr.io%2Fwebsocket+nats&type=code

A ton of devs are using the 2 together to solve the same problems it seems....

So this  example shows the concept: https://github.com/Mahamed-Belkheir/scalechat-backend#socket

The CMD is here: https://github.com/Mahamed-Belkheir/scalechat-backend/blob/master/socket_service/cmd/web/main.go#L16
- notice the NATS handel is passed into the Broker which handles the relationship between WS and NATS

https://github.com/Mahamed-Belkheir/scalechat-backend/blob/master/socket_service/cmd/web/main.go#L20
- the app is newed up, passing in the Broker.

https://github.com/Mahamed-Belkheir/scalechat-backend/tree/master/socket_service/api is the WS Service.

https://github.com/Mahamed-Belkheir/scalechat-backend/blob/master/socket_service/pubsub/ is the NATS Service

https://github.com/Mahamed-Belkheir/scalechat-backend/blob/master/socket_service/broker/broker.go#L41 looks to be the glue between the WS and NATS Services.

I only had like 5 minutes to dig into this example and did not clone and run it.

But i think this helps ?

**Comment by gedw99** [comment]
Centrifuge has a nice example too using the same imports

https://github.com/centrifugal/centrifuge/blob/master/_examples/custom_broker_nats/main.go

YOu can see the WS and NATS aspects i think...

The thing i like about Centrifuge is that is also handles presence: https://github.com/centrifugal/centrifuge/blob/master/_examples/custom_broker_nats/main.go#L110
- so you can tell if a client is connected and who it is always
